google/cloud/spanner/connection_options.h

```diff
@@ -152,13 +155,31 @@ class ConnectionOptions {
    */
   grpc::ChannelArguments CreateChannelArguments() const;
 
+  /**
+   * Configure the connection to use @p cq for all background work.
+   *
+   * Connections need to perform background work on behalf of the application.
+   * Normally they just create a background thread and a `CompletionQueue` for
+   * this work, but the application may need more fine-grained control of their
+   * threads. In this case the application can provide the
+   */
+  ConnectionOptions& DisableBackgroundThreads(
+      google::cloud::grpc_utils::CompletionQueue const& cq);
+
+  using BackgroundThreadsFactory =
+      std::function<std::unique_ptr<BackgroundThreads>()>;
+  BackgroundThreadsFactory background_threads_factory() {
```

Comment:
I'm not sure I understand how to use this method. It looks like every call to this getter will return a factory that, when invoked, will create a new "BackgroundThread", and by default, will use a new CompletionQueue. Is that right?

Would it be good to just expose access to a single CompletionQueue instance that everyone can use? But honestly, maybe I'm just not understanding how this works.

Reply:
> I'm not sure I understand how to use this method. It looks like every call to this getter will return a factory that, when invoked, will create a new "BackgroundThread", and by default, will use a new CompletionQueue. Is that right?

Yes.

> Would it be good to just expose access to a single CompletionQueue instance that everyone can use? But honestly, maybe I'm just not understanding how this works.

I think we decided to tie the lifetime of the automatically created threads to the lifetime of the `spanner::*Connection` object. That means the `*Connection` needs to know if it creates a thread and completion queue, and if it does create them it needs to shutdown the completion queue and join the thread in its destructor.  All those ifs suggested that this was a strategy given to the `*Connection` object.

In other words, we need to return something more than just the `CompletionQueue`, we need either (a) a thing that creates (when needed) the background thread for you, or (b) a `bool` that tells you if you should create a thread.

I went with option (a) because it seems wasteful to have every `spanner::*Connection` implement the "If I was asked to create a thread do this, otherwise do that".